### PR TITLE
Modernize Makefile

### PR DESCRIPTION
**Pull request by timkenhan**
related: https://github.com/gentoo/gentoo/pull/35089

### CODE REVIEW

**Comment by timkenhan** [comment]
@robbat2 can you help review?

File: Makefile [hunk 1]
```diff
@@ -5,27 +5,61 @@ MANPAGE = genkernel.8
...
+	gen_moddeps.sh \
+	gen_package.sh \
+	gen_worker.sh \
+	path_expander.py
+
+SOFTWARE = BCACHE_TOOLS BOOST BTRFS_PROGS BUSYBOX COREUTILS CRYPTSETUP \
+	DMRAID DROPBEAR EUDEV EXPAT E2FSPROGS FUSE GPG \
+	HWIDS ISCSI JSON_C KMOD LIBAIO LIBGCRYPT LIBGPGERROR LIBXCRYPT LVM \
+	LZO MDADM MULTIPATH_TOOLS POPT STRACE THIN_PROVISIONING_TOOLS UNIONFS_FUSE \
+	USERSPACE_RCU UTIL_LINUX XFSPROGS XZ ZLIB ZSTD
```
Comment (robbat2) [inline Makefile]: one entry per line please, to ease future diffs

File: Makefile [hunk 2]
```diff
@@ -5,27 +5,61 @@ MANPAGE = genkernel.8
...
+	gen_determineargs.sh \
+	gen_arch.sh \
+	gen_bootloader.sh \
+	gen_compile.sh \
+	gen_configkernel.sh \
+	gen_funcs.sh \
+	gen_moddeps.sh \
+	gen_package.sh \
+	gen_worker.sh \
+	path_expander.py
```
Comment (robbat2) [inline Makefile]: sort this list if the order is not signifcant. If it is significant, document why

File: Makefile [hunk 3]
```diff
@@ -56,9 +90,9 @@ distclean: clean
 		perl merge.pl $< $(BASE_KCONF) | sort > $@ ; \
 	fi ;
 
-%.8: doc/%.8.txt doc/asciidoc.conf Makefile genkernel
-	a2x --conf-file=doc/asciidoc.conf --attribute="genkernelversion=$(PACKAGE_VERSION)" \
```
Comment (robbat2) [inline Makefile:60]: i don't see the genkernel version being injected anywhere now; missing from this PR?

File: Makefile [hunk 4]
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
...
+
+	install -d $(DESTDIR)/$(PREFIX)/share/genkernel
+
+	cp -rp arch $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp defaults $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp gkbuilds $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp modules $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp netboot $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp patches $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp worker_modules $(DESTDIR)/$(PREFIX)/share/genkernel/
```
Comment (robbat2) [inline Makefile]:
```
DIRS = arch defaults gkbuilds modules netboot patches worker_modules

install: all
cp -ra $(DIRS) $(DESTDIR)/$(PREFIX)/share/genkernel/
```

File: Makefile [hunk 5]
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
...
+$(BUILD_DIR)/genkernel: $(addprefix $(BUILD_DIR)/,$(FINAL_DEPS)) $(BUILD_DIR)/software.sh
+	cp genkernel $(BUILD_DIR)/genkernel
+
+install: PREFIX := $(file <$(BUILD_DIR)/PREFIX)
+install: BINDIR := $(file <$(BUILD_DIR)/BINDIR)
+install: SYSCONFDIR := $(file <$(BUILD_DIR)/SYSCONFDIR)
+install: MANDIR := $(file <$(BUILD_DIR)/MANDIR)
+install: all
+	install -d $(DESTDIR)/$(SYSCONFDIR)
+	install -m 644 $(BUILD_DIR)/genkernel.conf $(DESTDIR)/$(SYSCONFDIR)/
```
Comment (robbat2) [inline Makefile:206]: why not `install -D $src $dest` ?

File: Makefile [hunk 6]
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
...
+	install -d $(BUILD_DIR)/
+	cp $< $@
+
+$(BUILD_DIR)/genkernel: $(addprefix $(BUILD_DIR)/,$(FINAL_DEPS)) $(BUILD_DIR)/software.sh
+	cp genkernel $(BUILD_DIR)/genkernel
+
+install: PREFIX := $(file <$(BUILD_DIR)/PREFIX)
+install: BINDIR := $(file <$(BUILD_DIR)/BINDIR)
+install: SYSCONFDIR := $(file <$(BUILD_DIR)/SYSCONFDIR)
+install: MANDIR := $(file <$(BUILD_DIR)/MANDIR)
```
Comment (robbat2) [inline Makefile]: is there a timing issue here? when are these rules evaluated? can you make sure it's always after the files are created? maybe throw error checking here to ensure they are non-empty.

File: Makefile [hunk 7]
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
...
+	touch $(BUILD_DIR)/build-config
+
+$(BUILD_DIR)/software.sh:
+	install -d $(BUILD_DIR)/temp/
+	echo -e $(SOFTWARE_VERSION) > $(BUILD_DIR)/temp/versions
+	cat $(BUILD_DIR)/temp/versions defaults/software.sh > $(BUILD_DIR)/software.sh
+
+$(BUILD_DIR)/doc/genkernel.8.txt:
+	install -d $(BUILD_DIR)/doc/
+	cp doc/genkernel.8.txt $(BUILD_DIR)/doc/genkernel.8.txt
```
Comment (robbat2) [inline Makefile]: you seem to mix `cp` and `install` a lot; this can be probably be make more consistent.

File: Makefile [hunk 8]
```diff
@@ -5,27 +5,61 @@ MANPAGE = genkernel.8
 # Add off-Git/generated files here that need to be shipped with releases
 EXTRA_DIST = $(MANPAGE) ChangeLog $(KCONF)
 
-default: kconfig man
```
Comment (robbat2) [inline Makefile:8]: keep; likely change to:
`default: all`
For non-GNU make engines if they are used

File: Makefile (hunk 2 above)
Comment (timkenhan) [reply Makefile]: addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

File: Makefile (hunk 1 above)
Comment (timkenhan) [reply Makefile]: addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

File: Makefile (hunk 4 above)
Comment (timkenhan) [reply Makefile]: thanks!

addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

File: Makefile (hunk 7 above)
Comment (timkenhan) [reply Makefile]: addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

File: Makefile (hunk 8 above)
Comment (timkenhan) [reply Makefile:8]: addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

File: Makefile (hunk 5 above)
Comment (timkenhan) [reply Makefile:206]: addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

File: Makefile (hunk 1 above)
Comment (robbat2) [reply Makefile]: now it's all on a single line? I expected
```
SOFTWARE = \
  BCACHE_TOOLS \
  BOOST \
  ...
```

File: Makefile (hunk 2 above)
Comment (robbat2) [reply Makefile]: thanks

**Review by robbat2 — COMMENTED** [review]
close; just those 2 last issues to clean up and then we can merge

File: Makefile (hunk 3 above)
Comment (timkenhan) [reply Makefile:60]: I removed this due to issue of not having git repo when using dist pkg.

I can add a code that generates a file with version upon `make dist`. The manpage generation code can then check for this file as well as the git to determine the version to inject. 

Any preference on the file path?

File: Makefile (hunk 1 above)
Comment (timkenhan) [reply Makefile]: lmao so sorry i completely misinterpreted your comment

addressed in commit [0050718](https://github.com/gentoo/genkernel/pull/56/commits/0050718347eb34b7fe36fc3cf421abf8dfa775b9)

File: Makefile (hunk 3 above)
Comment (timkenhan) [reply Makefile:60]: addressed in commit [c295e6b](https://github.com/gentoo/genkernel/pull/56/commits/c295e6b7ce4efa23d5b797e24c245e360a85c03b

lemme know if you think it needs to be done differently

File: Makefile (hunk 6 above)
Comment (timkenhan) [reply Makefile]: addressed in commit [ee3045f](https://github.com/gentoo/genkernel/pull/56/commits/ee3045fe2dca73560e685d535d9a14220bee7f20)

**Comment by timkenhan** [comment]
@robbat2 that should be everything! :)

**Review by robbat2 — APPROVED** [review]

**Comment by robbat2** [comment]
@timkenhan  it looks good now; rebase so it doesn't contain a merge and then i'll add it onto master

**Comment by timkenhan** [comment]
@robbat2 done!

**Comment by timkenhan** [comment]
@robbat2 also check out this PR on gentoo repo reflecting the recent changes:

https://github.com/gentoo/gentoo/pull/35089

**Comment by robbat2** [comment]
@timkenhan  somehow  your rebase picked up other commits that are already on the master branch?
```
gen_configkernel: Prepare for kernel-6.6 
doc: add new cmdline options to man page 
```
I'm going to exclude them for now when I merge this

**Comment by robbat2** [comment]
For future commits, please also ensure you have signed-off-by trailers on each commit.

**Pull request closed without merging**